[2.2 regression] unreachable code regression with custom equality enum membership check

**Bug Report**

Mypy 2.2.0 introduced a regression from 2.1.0 with equality comparisons with custom integer enums. 

Using the config option `warn_unreachable = True`, the following code checks successfully in mypy 2.1.0, but under mypy 2.2.0 fails with 
```
test.py:23: error: Statement is unreachable  [unreachable]
```
despite the code actually hitting this path at runtime and printing "Success". 


**To Reproduce**

```python
#!/usr/bin/env python3
import enum
import sys
import typing

class E(enum.Enum):
    foo = 1
    bar = 2

    def __int__(self) -> int:
        return int(self.value)

    def __eq__(self, other: typing.Any) -> bool:
        return bool(int(self) == other)

class Test:
    def __init__(self) -> None:
        self.val: int  = 1

t = Test()

if t.val in [E.foo, E.bar]:
    print("success")
    sys.exit(0)
print("fail")
sys.exit(1)
```

**Your Environment**

- Mypy version used: 2.2.0
- Mypy command-line flags: `mypy test.py` 
- Mypy configuration options from `mypy.ini` (and other config files): `warn_unreachable = True`
- Python version used: 3.11

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -6858,9 +6858,13 @@ class TypeChecker(NodeVisitor[None], TypeCheckerSharedApi, SplittingVisitor):
                         for known_item in container_item_types:
                             # Match the should_coerce_literals logic from narrow_type_by_identity_equality
                             p_known_item = get_proper_type(known_item)
-                            if is_literal_type_like(p_known_item) or (
-                                isinstance(p_known_item, Instance) and p_known_item.type.is_enum
-                            ):
+                            if (
+                                is_literal_type_like(p_known_item)
+                                or (
+                                    isinstance(p_known_item, Instance)
+                                    and p_known_item.type.is_enum
+                                )
+                            ) and not has_custom_eq_checks(p_known_item):
                                 known_item = coerce_to_literal(known_item)
                             if_map, else_map = self.narrow_type_by_identity_equality(
                                 "==",

--- a/test-data/unit/check-narrowing.test
+++ b/test-data/unit/check-narrowing.test
@@ -3335,6 +3335,24 @@ def narrow_dict(x: Literal['a', 'b', 'c'], t: dict[Literal['a', 'b'], int]):
 [builtins fixtures/primitives.pyi]
 
 
+[case testNarrowCustomEqEnumInLiteralContainer]
+# flags: --strict-equality --warn-unreachable
+# https://github.com/python/mypy/issues/21703
+from enum import Enum
+
+class E(Enum):
+    foo = 1
+    bar = 2
+
+    def __eq__(self, other: object) -> bool:
+        return True
+
+def f(x: int) -> None:
+    if x in [E.foo, E.bar]:
+        reveal_type(x)  # N: Revealed type is "builtins.int"
+[builtins fixtures/list.pyi]
+
+
 [case testNarrowingLiteralInLiteralContainer]
 # flags: --strict-equality --warn-unreachable
 from typing import Literal

